co-log/src/Colog/Concurrent.hs

```diff
@@ -0,0 +1,129 @@
+-- |
+-- Sometimes an application may need to dump logs asynchonously.
+-- This may happen if the target where logs are dumped is
+-- slow and logs should be dumped in chunks, or if application
+-- should split CPU and I/O-bound tasks. In these cases it
+-- may make sense to run asynchronous logger.
+--
+-- Asynchronous logger runs in background thread, so users
+-- thread may dump logs asynchronously.
+--
+-- This module provides basic interface for the concurrent
+-- logging that may be either used as-is or can be used to
+-- build more complex concurrency abstractions.
+module Colog.Concurrent
+       ( -- $worker-thread
+         mkBackgroundThread
+       , runInBackgroundThread
+       -- $background-worker
+       , BackgroundWorker
+       , forkBackgroundLogger
+       , killBackgroundLogger
+       , convertToLogAction
+       ) where
+
+import Control.Concurrent
+import Control.Concurrent.STM.TBQueue
+
+import Colog.Core.Action (LogAction (..))
+import Colog.Concurrent.Internal
+
+-- $worker-thread
+-- Worker thread approach allows to create a single worker
+-- thread that will perform all logging. This approach allows
+-- user to use different loggers and change them on the flight.
+--
+-- With bracket resource control style code may look as
+-- following:
+--
+-- @
+-- bracket ('mkBackgroundThread' 4096)
+--         ('killBackgroundThread')
+--         $ \thread -> 
+--            let la1 = runInBackgroundThread worker logger1
+--                la2 = runInBackgroundThread worker logger1
+--            unLogger la1 msg1
+--            unLogger la2 msg2
+--            ..
+-- @
+--
+-- This code will run a background thread and process different
+-- loggers in that thread.
+--
+-- You may extend the logger and entire logger will be run
+-- on the background thread. It is not a case with
+-- 'forkBackgroundLogger'.
+
+-- | Create a background worker with a given capacity.
+-- If capacity is reached, then the thread that tries to
+-- write logs will be blocked.
+mkBackgroundThread :: Int -> IO (BackgroundWorker (IO ()))
+mkBackgroundThread capacity = do
+  queue <- newTBQueueIO (fromIntegral capacity)
+  tid <- forkIO $ forever $ join $ atomically $ readTBQueue queue
+  pure $ BackgroundWorker tid (writeTBQueue queue)
+
+-- | Run logger action asynchronously in the worker thread.
+-- Logger is executed in the other thread entirely, so if
+-- logger takes any thread related context it will be
+-- read from the other thread.
+runInBackgroundThread :: BackgroundWorker (IO ()) -> LogAction IO msg -> LogAction IO msg
```

Comment:
It's possible to have the result type of `runInBackgroundThread` function to be `MonadIO m => LogAction m msg` but I wonder how useful it can be. Whole current API is built on top of the plain `IO` and the only sensible generalisation that I know involves `MonadUnliftIO` but I'm not sure it's a good idea to bring `MonadUnliftIO` now.

Reply:
It can be useful, because that way you may provide useful capabilities there. But I'm uneasy with that change, currently, there are a number of different approaches that allow such generalizations:

1. MonadUnliftIO and unliftio ecosystem in general
2. monad-control that is still widely used
3. MTL-heavy libraries that are using MTL as final tagless system
4. a family of the effects systems.
5. direct handle approach

Making a preference to the one system other the others may be a bad thing, as it increases dependencies set and makes people used to other systems uneasy when using your library.

And `IO` on it's own is very powerful and composable type, in general with any of the approaches mentioned above you can write `LogAction IO msg` and "unlift" your system to those. So basically you need `unlift :: SomeConstraints m => LogAction m msg -> LogAction IO msg` that will be called in the parent thread.

Reply:
Okay, I can just add `liftLogIO :: MonadIO m => LogAction IO msg -> LogAction m msg` to `co-log-core` to make life of users easier. So they have some way accessible without extra dependencies introduced.